IPIP-296: DNSLink resolution export with DNSSEC records (#296)
ipfs/specs: http-gateways/DNSSEC_EXPORT_DNSLINK.md

- Start Date: 2022-07-06
- Related Issues:
  - https://github.com/ipfs/kubo/issues/8799
  - https://github.com/ipfs/kubo/issues/6129

## Summary
IPFS is connected to DNS through a standard called DNSLink. DNSLink allows a DNS TXT record for a domain to point to an IPNS address or IPFS content ID (CID). This specification describes a method for providing proof to the requesting client that the DNS mapping is provided by honest/trusted name servers and that the IPFS address pointed to by the TXT record corresponds to the content received by the client.

This specification consists of four logical pieces.

1. A method for serializing DNSSEC proofs and IPFS CAR files
2. A method for communicating these proofs to the requesting client
3. A method for the client to validate DNSSEC proofs
4. A method for the client to validate the integrity of the IPFS content

## Motivation
In the current gateway, clients are forced to trust the DNS and IPNS resolutions of the gateway. This entails implicit risk for clients who cannot validate if the content they receive is the content bound to the name they requested. This state of affairs can leave clients vulnerable to MitM attacks.

> This proposal reduces the trust needed by a client, and allows honest gateways to prove it is responding truthfully.

Reviewer: Wouldn't "... , and allows an honest gateway to prove it is responding truthfully."

or "... , and allows honest gateways to prove they are responding truthfully." be more grammatically correct?

> This proposal reduces the trust needed by a client, and allows honest gateways to prove it is responding truthfully.

Author: ```suggestion
This proposal reduces the trust needed by a client, and allows honest gateways to prove they are responding truthfully.
```

2. Gateways should maintain compatibility with non-participating clients—those clients will not benefit from the DNSSEC proofs, but they should still have full access to data.
3. Gateways should not rely on any specific public DNS resolver, i.e., the implementation should be resolver agnostic.

### Security Considerations
1. DNSSEC responses can be much larger that vanilla DNS responses, in theory, this could be a vector for DDoS amplification.

### Assumptions on Scope
1. We consider only IPFS blocks bound to DNS with DNSLink. This is without loss of generality, because an IPFS subgraph referenced by its root hash is directly verifiable by confirming the hashes and thus would not benefit from the DNSSEC proof.
2. For non-existent URLs, we will not walk the zone (through NSEC records). Including the walkthrough would dramatically increase the proof size. In other words, we will only offer verifiable proof for positive responses.
3. We will not support wildcards in DNS resolution. This is not a problem because our target subdomains will always begin with "\_dnslink", and domains cannot have wildcards in the middle.

> 4. We will not support multiple root Key Signing Keys or changed Key Signing Keys in this version.

Author: I wrote that because if the root KSK changes, then whatever client-side program is doing the DNSSEC validation would need to be updated out of band to use the new KSK.

But you're right that I generally assumed that the client is aware of the current KSK situation. I think it will be more clear if I remove \#4 and replace it with an explicit statement about the client's awareness of the root's KSK situation.

```suggestion
4. We assume that the client is configured to use the correct root zone KSK.
```

- A gateway can attempt provide either a false DNSSEC proof or incorrect IPFS content and is therefore untrusted.

## Detailed Design
### Components
1. HTTPS-capable client (likely in a browser)
2. IPFS gateway
3. Recursive resolver
4. IPFS network

### How do the components work together
1. Client establishes a TLS connection to the gateway and sends a HTTP request to the IPFS gateway with header "Accept: application/vnd.ipld.namedcar" using a Reduced TLS DNSSEC Chain Extension (see below)
2. Gateway queries recursive resolver for DNSSEC chain and DNS resolution, which should yield an IPFS content name (from a DNSLink)
3. Gateway requests CAR of content from IPFS network
4. Gateway returns IpfsDnssecChain (see data model) to client and also caches for future requests.

```mermaid
stateDiagram-v2
    state "IPFS Network" as ipfs
    Client --> Gateway: HTTP GET\n cloudflare-ifps.io/ipfs.example.com
    Gateway --> Client: IpfsDnssecChain
    Gateway --> Resolver: DNSSEC query\n ipfs.example.com
    Resolver --> Gateway: DNSSEC response\n /ipns/abc123
    Gateway --> ipfs: ipfs get /ipns/abc123
    ipfs --> Gateway: IPFS file
```
            

> ### Reduced TLS DNSSEC Chain Extension

Reviewer 2: The Reduced TLS DNSSEC Chain Extension is an experimental RFC, which has inspired some of the design in this IPIP. There is no ask for gateways to implement a custom TLS layer.

Validation begins in the INITIAL state. Upon reading the CAR header, we extract the CIDs and payloads of each block in the CAR (stored in a list), then enter the VALIDATE\_HASH state. In VALIDATE\_HASH, we check that the client has support for the hash formats and codecs of the current block (and fail if not). We then hash the block payload and check that it matches the CID. If they do not match, validation fails. If they do match, we move to the next block in the list and reenter the VALIDATE\_HASH state. If there are no more blocks, transition to ACCEPT.

```mermaid
stateDiagram-v2
    INITIAL --> VALIDATE_HASH: Parse blocks, [(CID, payload)]
    INITIAL --> FAIL: Cannot parse CAR
    VALIDATE_HASH --> VALIDATE_HASH: CID == hash(payload)
    VALIDATE_HASH --> FAIL: CID != hash(payload)
    VALIDATE_HASH --> ACCEPT
```
For the client implementation, we will prefer to use an available library. Both Go and JS libraries are available (i.e., go-car and js-car). If we opt for the browser extension design, the JS library is probably the best option.

### Alternative Design

> For each request, the gateway would respond with an additional header of the form X-Ipfs-Name-Proof: <CID>, where <CID> is a pointer to an IPFS file containing the DNSSEC proof. A client could then retrieve this proof directly from IPFS if they want it. However, we decided against this approach for several reasons. The main reason is that it is never more advantageous for the gateway to retrieve the DNSSEC proof from IPFS than it would be to get it from a local cache. The only time this would be useful is if a gateway could use DNSSEC proofs found by other gateways. This is not feasible because the DNSSEC proof would be stored according to its hash (in IPFS), and the hash required can only be known by already collecting the DNSSEC proof—therefore it would be pointless for a gateway to request content that, by definition, it already has. An additional problem is that DNSSEC proofs have TTLs, and thus will not usually be valid for more than an hour. This limits the time that the content would be useful anyway.

Reviewer 2: having the resolution in two steps require the gateway to keep some state, which I don't think is a good requirement. Ideally, gateway should be stateless. In the variant you propose above, a stateless gateway could be consider dishonest as it might not necessarily have the same view of DNS in the two requests.

For a two requests protocol to not require the gateway to have state, a client should first request the DNSSEC proof, and then request the resolved CID/CAR.

> For each request, the gateway would respond with an additional header of the form X-Ipfs-Name-Proof: <CID>, where <CID> is a pointer to an IPFS file containing the DNSSEC proof. A client could then retrieve this proof directly from IPFS if they want it. However, we decided against this approach for several reasons. The main reason is that it is never more advantageous for the gateway to retrieve the DNSSEC proof from IPFS than it would be to get it from a local cache. The only time this would be useful is if a gateway could use DNSSEC proofs found by other gateways. This is not feasible because the DNSSEC proof would be stored according to its hash (in IPFS), and the hash required can only be known by already collecting the DNSSEC proof—therefore it would be pointless for a gateway to request content that, by definition, it already has. An additional problem is that DNSSEC proofs have TTLs, and thus will not usually be valid for more than an hour. This limits the time that the content would be useful anyway.

Author: In addition, the gateway doesn't need to request the proofs from the resolver unless it receives the header `"Accept: application/vnd.dnslink.dnssec; version=1"`. So a client who's only interested in the CAR can opt to receive only that.

> For each request, the gateway would respond with an additional header of the form X-Ipfs-Name-Proof: <CID>, where <CID> is a pointer to an IPFS file containing the DNSSEC proof. A client could then retrieve this proof directly from IPFS if they want it. However, we decided against this approach for several reasons. The main reason is that it is never more advantageous for the gateway to retrieve the DNSSEC proof from IPFS than it would be to get it from a local cache. The only time this would be useful is if a gateway could use DNSSEC proofs found by other gateways. This is not feasible because the DNSSEC proof would be stored according to its hash (in IPFS), and the hash required can only be known by already collecting the DNSSEC proof—therefore it would be pointless for a gateway to request content that, by definition, it already has. An additional problem is that DNSSEC proofs have TTLs, and thus will not usually be valid for more than an hour. This limits the time that the content would be useful anyway.

Reviewer 3: I agree on two points: the gateway should be stateless, and the client should be able to decide when the DNSLink proof is returned.

What I'm not comfortable with is limiting clients, forcing them to always fetch data when all they want is DNSLink proofs.  Having 2-in-1 response type is fine (sounds like you have use cases in mind for this?), but there should also be a way to request proof without the data, allowing clients to do the data retrieval by other means.

> For a two requests protocol to not require the gateway to have state, a client should first request the DNSSEC proof, and then request the resolved CID/CAR.

This is exactly that I had in mind, and it follows the patterns of how the web works.

You need to resolve DNSLink once, and in my mind DNSSEC proof would be part of that.
After that, you don't need it until TTL runs out.

## Data Model
In this section we define the data model and structures that define the serialization format. All data will be serialized in the order specified in the following structs, without padding. All multi-byte integers are encoded big-endian. The top-level structure is IpfsDnssecChain, all data necessary for a client to validate the resolution's correctness is nested within this structure.

```c
struct IpfsDnssecChain {
  uint8 version;
  DnssecProof proof;
  Car ipfsData;

> }

Reviewer 2: coupling has the advantage of not requiring state, nor a two requests protocol from the client side. Similar to CAR which block headers could technically be requested separately, having them all bundled in one file is useful.
* It allows the state to be consistent when generating the bundle, and when validating it.
* Caching is also simpler as there is only one DNS/IPNS/etc view that prevails: the gateway view.

Regarding the decoupling, one could imagine a DAG build on top of IpfsDnssecChain with a node for version, a node for proof, and a node for ipfsData. Data retrieval is different from the bundling. One will likely not be able to resume download of a DNSLink proof on a gateway with a different view, but once the bundle has been generated, it can be content addressed.

> }

Reviewer 3: > One will likely not be able to resume download of a DNSLink proof on a gateway with a different view

Why not? Client resolved DNSLink to a CID.  CID is either a single block, or a root of a bigger DAG (e.g., UnixFS directory,).
 
If that CID is a root of a DAG, download can be not only resumed, but download of different branches of the DAG can be fetched from different gateways, or even via P2P transports.

```

> Note that simply including both the DNSSEC chain and the IPFS CAR in the response is sufficiently strong binding for this use case. This is because the last entry in the DNSSEC chain will include a signed DNSLink record, which validates the IPFS root hash, meanwhile, the client can directly confirm that that CAR is correct by hashing.

Author: I think @thibmeu's comments apply here as well. If the binding fails to validate, it would not be clear if the gateway was dishonest or if the DNSLink mapping changed.

  RRSIG = 2,
  DS = 3,
  CNAME = 4,
  DNAME = 5,
  TXT = 6
}
 
struct RRData {
  uint16 length;
  byte rrdata[length];
}
```

Keys are formatted exactly as in RFC 4034 (Section 2.1), and ordered as in Section 6.3.

```c
struct Key {
  uint16 length;
  Dnskey_Rdata rdata[length];
}
 
struct Dnskey_Rdata {
  uint16 flags;
  uint8 protocol;
  uinit8 algorithm;
  byte public_key[];
}
```

This struct represents the signature and associated metadata.

```c
struct Signature {
  uint16 length;
  uint8 algorithm;
  uint8 labels;
  uint32 ttl;
  uint32 expires;
  uint32 begins;
  uint16 key_tag;
  byte signature[length - 16];
}
```

This struct holds the data needed to validate DS records. This format is the same as described in RFC 4034 (Section 5.1). According to [Serializing DNS Records with DNSSEC Authentication](https://www.ietf.org/archive/id/draft-agl-dane-serializechain-01.html#rfc.section.3), key\_tag and algorithm are redundant. However, inferring both of these increases complexity, so we will include them explicitly in version 1.

```c
struct Ds {
  uint16 key_tag;
  uint8 algorithm;
  uint8 digest_type;
  uint16 digest_len;
  byte digest[digest_len];
}
```

### Security 

> DDoS Amplification: For each request to the gateway, a request must be made to the resolver and to IPFS, leading to several server-side computations. However, This information is cacheable, which should reduce the risk.

Reviewer 2: Requiring a minimum TTL could go against DNS behaviour, and adding a new layer of cache should be left to the gateway operator.

> DDoS Amplification: For each request to the gateway, a request must be made to the resolver and to IPFS, leading to several server-side computations. However, This information is cacheable, which should reduce the risk.

Reviewer 3: Right, I forgot we already talked about his around  [`DNS.MaxCacheTTL`](https://github.com/ipfs/kubo/blob/master/docs/config.md#dnsmaxcachettl).  

Ignore my comment then, no special treatment here, 
gateway and client should reuse DNS cache around TTL value for TXT record with DNSLink.
